ludwig/models/modules/sequence_decoders.py

```diff
@@ -15,134 +15,413 @@
+        end_token = END_SYMBOL
+
+        decoder_emb_inp = self.decoder_embedding(decoder_input)
+
+        # code sequence based on example found here
+        # https://www.tensorflow.org/addons/api_docs/python/tfa/seq2seq/BeamSearchDecoder
+        tiled_encoder_output = tfa.seq2seq.tile_batch(
+            encoder_output,
+            multiplier=self.beam_width
+        )
+
+        tiled_encoder_end_state = tfa.seq2seq.tile_batch(
+            encoder_end_state,
+            multiplier=self.beam_width
+        )
+
+        tiled_encoder_sequence_length = tfa.seq2seq.tile_batch(
+            encoder_sequence_length,
+            multiplier=self.beam_width
+        )
+
+        if self.attention_mechanism is not None:
+            self.attention_mechanism.setup_memory(
+                tiled_encoder_output,
+                memory_sequence_length=tiled_encoder_sequence_length
+            )
+
+        decoder_initial_state = self.build_decoder_initial_state(
+            batch_size * self.beam_width,
+            encoder_state=tiled_encoder_end_state,
+            dtype=tf.float32
+        )
+
+        decoder = tfa.seq2seq.beam_search_decoder.BeamSearchDecoder(
+            cell=self.decoder_rnncell,
             beam_width=self.beam_width,
-            num_layers=self.num_layers,
-            attention_mechanism=self.attention_mechanism,
-            is_timeseries=self.is_timeseries,
-            embeddings=tied_embeddings_tensor,
-            initializer=self.initializer,
-            regularizer=regularizer
+            output_layer=self.dense_layer)
+        # ================generate logits ==================
+        maximum_iterations = self.max_sequence_length
+
+        # initialize inference decoder
+        decoder_embedding_matrix = self.decoder_embedding.variables[0]
+        (
+            first_finished,
+            first_inputs,
+            first_state
+        ) = decoder.initialize(
+            decoder_embedding_matrix,
+            start_tokens=start_tokens,
+            end_token=end_token,
+            # following construct required to work around inconsistent handling
+            # of encoder_end_state by tfa
+            initial_state=decoder_initial_state \
+                if len(decoder_initial_state) != 1 else decoder_initial_state[0]
+        )
+
+        inputs = first_inputs
+        state = first_state
+
+        # create empty prediction tensor
+        predictions = tf.convert_to_tensor(
+            np.array([]).reshape([batch_size, 0]),
+            dtype=tf.int32   # todo tf2 need to change to tf.int64
+        )
+
+        # beam search
+        for j in range(maximum_iterations):
+            outputs, next_state, next_inputs, finished = decoder.step(
+                j, inputs, state, training=training)
+            inputs = next_inputs
+            state = next_state
+            one_predicted_token = tf.expand_dims(outputs.predicted_ids[:,0], axis=1)
```

Comment:
this seems kinda weird to me: At each step we are taking the best token and adding it to the predictions, but that's against the spirit of beam search. The best token at the current step may be the worst token when we rank sequences using the probabilities of the following tokens.

What we need to do is to run the beam search process until the end, and only then take the first most likely sequence adn return both logits and predictions.